src/ripple/rpc/impl/ShardArchiveHandler.cpp

```diff
@@ -61,24 +94,156 @@ ShardArchiveHandler::hasInstance()
     return instance_.get() != nullptr;
 }
 
-ShardArchiveHandler::ShardArchiveHandler(Application& app, bool validate)
+ShardArchiveHandler::ShardArchiveHandler(WithoutRecovery const& tag,
+                                         Application& app, bool validate)
     : app_(app)
-    , downloadDir_(get(app_.config().section(
-        ConfigSection::shardDatabase()), "download_path",
-            get(app_.config().section(ConfigSection::shardDatabase()),
-                "path", "") + "/download"))
+    , j_(app.journal("ShardArchiveHandler"))
+    , downloadDir_(getDownloadDirectory(app.config()))
+    , stateDir_(getStateDirectory(app.config()))
     , validate_(validate)
     , timer_(app_.getIOService())
     , process_(false)
-    , j_(app.journal("ShardArchiveHandler"))
 {
     assert(app_.getShardStore());
+
+    auto setup = setup_DatabaseCon(app_.config());
+    setup.dataDir = stateDir_;
+
+    create_directory(stateDir_);
+
+    conn_ = std::make_unique<DatabaseCon>(
+        stateDir_,
+        "stateDatabase",
+        DatabasePragma,
+        DatabaseInit);
+
+    auto db = conn_->checkoutDb();
+
+    // Always use named variables with 'soci::use'.
+    // Avoid temporaries.
+    int v = int(validate_);
+
+    *db << "INSERT INTO MetaState VALUES (\"validate\", :val);",
+        soci::use(v);
+
+    *db << "INSERT INTO MetaState VALUES (\"paused\", 0);";
+
+    (void)tag;
+}
+
+ShardArchiveHandler::ShardArchiveHandler(WithRecovery const& tag,
+                                         Application& app, bool & invokeStart)
+    : app_(app)
+    , j_(app.journal("ShardArchiveHandler"))
+    , downloader_(std::make_shared<DatabaseDownloader>(
+        app_.getIOService(), j_, app_.config()))
+    , downloadDir_(getDownloadDirectory(app.config()))
+    , stateDir_(getStateDirectory(app.config()))
+    , timer_(app_.getIOService())
+    , process_(false)
+{
+    auto setup = setup_DatabaseCon(app_.config());
+    setup.dataDir = stateDir_;
+
+    using namespace boost::filesystem;
+
+    assert(exists(stateDir_.string() + "/stateDatabase") &&
+           is_regular_file(stateDir_.string() + "/stateDatabase"));
+
+    conn_ = std::make_unique<DatabaseCon>(
+        stateDir_,
+        "stateDatabase",
+        DatabasePragma,
+        DatabaseInit);
+
+    auto db = conn_->checkoutDb();
```

Comment:
Same as previous comment. No need for a LockedSociSession. conn_ is always used while holding the lock, getSession() can be used instead.

Reply:
Done